---
title: "<span style='font-size: 20pt; font-weight: bold;'>scRNA-seq analysis of single species - <i>Cardamine hirsuta</i></span>"
subtitle: "<span style='font-weight: bold;'>Materials and Methods</span>"
subtitle_font_size: 18pt
author: "Yasir Arafat Tamal"
date: "2023-09-30"
output: 
  html_document:
        css: custom.css
        fig_height: 9
        fig_width: 11
        toc: true
        toc_float: false
        number_sections: true
editor_options: 
  chunk_output_type: inline
---

<style>
a:link {
    color: red;
}

a:visited {
    color: green;
}

 a:hover {
    color: hotpink;
}

</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The materials, methods, and scripts used in this chapter can be found in this GitHub repository:
(https://github.com/tamalarafat/Analysis_of_single_species_Cardamine)

# <span style='font-size: 16px; font-weight: bold;'>Plant materials and experimental design</span>
The _C. hirsuta_ plants were grown in the soil in a greenhouse under short-day conditions (18°C in 16 hours dark and 21°C in 8 hours light). Developing leaf primordia from leaves 5, 6, and 7 were collected three weeks post-germination. Four biological replicates were collected for the leaf samples to generate transcriptomic profiles at the single-cell level. All the experiments were carried out in our department (Department of Comparative Development and Genetics), with fellow colleagues Dr. Bazakos and Nora Papadima-Karanikou handling the experimental setup and execution.

# <span style='font-size: 16px; font-weight: bold;'>Bulk RNA-seq library preparation</span>
Bulk RNA-seq libraries were generated from equivalent protoplasted and un-protoplasted leaf tissues. Two replicates (one wild-type genotype and one transgenic line (rco-mutant)) for the protoplasted tissue and two replicates (one wild-type genotype and one transgenic line (rco-mutant)) for the un-protoplasted tissue were collected to generate bulk RNA-seq libraries. The bulk RNA-seq experiments were carried out in our department by Dr. Bazakos.

# <span style='font-size: 16px; font-weight: bold;'>Bulk RNA-seq analysis</span> {#brna_analysis}
Genes induced by protoplasting were identified (absolute fold-change >= 2; q < 0.05) by differential expression (DE) analysis between the protoplasted and un-protoplasted bulk RNA-seq experiments. Dr. Bazakos, conducted the comparative bulk RNA-seq analysis.

# <span style='font-size: 16px; font-weight: bold;'>Single cell library preparation</span>
Single-cell RNA-seq libraries were prepared from fresh protoplasts according to the 10x Genomics Single cell 3' Reagent Kit v2 protocol (Zheng et al., 2017). The standard Cell Ranger pipeline has been used to generate cellular gene expression profiles (Zheng et al., 2017). These libraries were prepared by fellow colleagues Dr. Bazakos and Nora Papadima-Karanikou.

# <span style='font-size: 16px; font-weight: bold;'>Single cell expression matrices</span>
Cellular gene expression profiles have been generated from four biological replicates. The scRNA-seq datasets of _C. hirsuta_ contain gene expression profiles of 4100, 6640, 9090, and 10760 cells with an average median of 5195 genes and an average median of 21,637 unique molecular identifier (UMI) counts per cell. The _C. hirsuta_ reference genome contains a total of 27,773 genes.

# <span style='font-size: 16px; font-weight: bold;'>Literature-derived markers lists</span>
Nora Papadima-Karanikou conducted a bibliographical search and compiled a list of cell-type markers for important leaf structures like the epidermis, mesophyll, and vasculature of _A. thaliana_. In a similar search, Dr. Bazakos created a list of known cell cycle genes specific to _A. thaliana_.

# <span style='font-size: 16px; font-weight: bold;'>Tools and software</span>
The R programming language (R Core Team, 2022) was used throughout the thesis to perform each data analysis task. This supplementary document lists all the methods, tools, and packages and their version that have been utilized to perform any data analysis task (Supplementary Table 8).

# <span style='font-size: 16px; font-weight: bold;'>Quality control and statistical analysis</span>

## <span style='font-size: 16px; font-weight: bold;'>Pre-processing of scRNA-seq data</span>
Each dataset was subjected to every pre-processing step. The pre-processing steps follow the order mentioned below:

### <span style='font-size: 16px; font-weight: bold;'>Filtering of the scRNA-seq data</span>
First, protoplasting-induced genes (identified from Bulk RNA-seq analysis; Section [Bulk RNA-seq analysis](#brna_analysis)) were removed. Second, genes for which expression was not detected in a minimum 0.2% cell population (1 out of 500 cells) were excluded. Third, cells expressing less than 200 genes or with counts larger than 110,000 were excluded. Fourth, cells were excluded if the proportion of the read count of mitochondrial genes was 5% or more and if the proportion of chloroplast genes was 10% or more. The _"PercentageFeatureSet"_ function of the _Seurat_ tool (Stuart et al., 2019) was used to calculate the percentage of counts originating from mitochondrial genes and chloroplast genes.

### <span style='font-size: 16px; font-weight: bold;'>Normalization of the data</span>
The scRNA-seq data was then log-transformed to normalize the expression count to account for differences in sequencing depth and efficiency between cells. The gene expression counts of each cell were divided by the total count accounted for that cell and then multiplied by a scale factor of 10000. The scaled counts of the cells were then subjected to natural-log transformation. The _"NormalizeData"_ function implemented in the _Seurat_ tool was used for the normalization task (Stuart et al., 2019).

### <span style='font-size: 16px; font-weight: bold;'>Selection of highly variable genes</span>
A set of highly variable genes was selected to reduce the dimensionality of the data and select a set of genes of biological relevance with cell type discrimination ability. A detailed explanation of the highly variable gene selection method has been described (Stuart et al., 2019). The _"FindVariableFeatures"_ function of the _Seurat_ tool was used to select a set of 2000 highly variable genes in the data.

### <span style='font-size: 16px; font-weight: bold;'>Selection of integration features</span>
Next, a set of genes, often referred to as the integration features, that were repeatedly selected as highly variable features across the datasets were selected to integrate the scRNA-seq datasets. The integration features selection method has been described in detail (Stuart et al., 2019). The _"SelectIntegrationFeatures"_ function of the _Seurat_ tool was used to select the integration features.

### <span style='font-size: 16px; font-weight: bold;'>Gene-level scaling</span>
Gene-level scaling was performed to scale the expression level of the genes across the cells to a common scale. This ensures that the expression level of the genes is comparable, and they contribute equally in the downstream analysis, such as clustering.

#### <span style='font-size: 16px; font-weight: bold;'>Unit variance scaling</span>
The log-transformed expression measurement of the genes across the cells was divided by the standard deviation to prepare data for the _Liger_ integration procedure. This results in a distribution of the expression values of a gene with a standard deviation of 1. The _"scaleNotCenter"_ function of the _Liger_ tool was used to apply unit variance scaling so that each gene has the same variance (Welch et al., 2019). Note that the NMF or NMF-based methods (_Liger_) require positive values in the input data, so we do not center the data by subtracting the mean. Gene-level scaling is done in the _Liger_ analysis workflow before integrating the datasets.

#### <span style='font-size: 16px; font-weight: bold;'>Zero mean and unit variance scaling</span> {#std_scaling}
_Seurat_ implements standardization by transforming the genes to have zero mean and unit variance. On the log-transformed count data, this is achieved by subtracting the mean value of a gene from each cell and dividing it by the standard deviation. The _"ScaleData"_ function of the _Seurat_ tool was used to perform the gene-level scaling task (Stuart et al., 2019). In the _Seurat_ analysis workflow, gene-level scaling is done after integrating the datasets but before performing clustering analysis.

## <span style='font-size: 16px; font-weight: bold;'>Non-negative matrix factorization (NMF)</span>
The non-negative matrix factorization (NMF) is a parts-based, low-rank matrix approximation technique (Brunet et al., 2003). It is widely used as an unsupervised dimension reduction method for biological pattern discovery and an efficient method for cluster discovery. The capacity to recognize localized structures that jointly interpret data as a linear mixture, subject to the requirement that only expression signatures can be added, never subtracted, is a primary characteristic of this approach (H. Kim & Park, 2007).

## <span style='font-size: 16px; font-weight: bold;'>Data integration</span>
Integrating scRNA-seq data of biological replicates, genotypes, or species serves different purposes in analyzing scRNA-seq data. Integrating data from multiple independent experiments on the same biological tissue or different tissue samples improves the robustness and reliability of analysis results by mitigating the batch effects and capturing biological signals more accurately (Stuart et al., 2019). The integration procedure combines independent experiments' scRNA-seq data into a shared space for downstream analysis. We used two integration techniques to perform integrative analysis between biological replicates of the same tissue and comparative analysis between experimental conditions (wild-type and gene-mutant genotype) or between species.

### <span style='font-size: 16px; font-weight: bold;'>_Liger_ integration procedure</span>
After performing the pre-processing steps on the individual datasets, all the datasets are pulled into a single object using the _"createLiger"_ function of the _Liger_ tool (J. Liu et al., 2020). _Liger_ uses an NMF-based framework, integrative non-negative matrix factorization (iNMF), to integrate the scRNA-seq datasets. The Macoslab developed a novel block coordinate descent algorithm for performing iNMF. This approach learns a set of latent factors, each with both shared and dataset-specific components, to approximate the original datasets. A detailed description of the method can be found (Welch et al., 2019). 

The _Liger_ integration procedure involves two steps (Tran et al., 2020; Welch et al., 2019). First, utilizing the iNMF, a joint factorization is performed on the input data (pulled object), and a low-dimensional representation of the data is obtained, representing the aligned datasets. This factorization step yields several lower-dimension matrices (components of the low-dimensional representation), including a dataset-specific basis matrix for each dataset (batch or dataset-specific part) that captures the dataset-specific components representing the batch-specific variation, a dataset-specific coefficient matrix for each dataset (GEP loadings for each cell), and a shared basis matrix for all datasets (a set of shared components) that captures the shared variation across the datasets. The joint factorization of the data was performed using the _"optimizeALS"_ function provided by the _Liger_ tool (J. Liu et al., 2020). In the next step, to correct for batch effects and identify groups of cells sharing the same developmental processes across the datasets, _Liger_ employs a shared factor neighborhood graph method to search for pairs of cells with similar transcriptomic profiles and then connect these pairs of cells with similar neighborhoods. All of these functionalities are encapsulated within the _"quantile_norm"_ function in the _Liger_ tool, which uses max factor assignment and refinement using a k-nearest neighbors graph (J. Liu et al., 2020). This step produces joint clustering assignments and a shared low-dimensional representation (shared coefficient matrix) of the combined data that integrates all the datasets into a shared space.

### <span style='font-size: 16px; font-weight: bold;'>_Seurat_ integration procedure</span>
The _Seurat_ integration approach also involves two steps. In the first step, _Seurat_ identifies correlations across the datasets by projecting cells of all datasets into a low-dimensional space using canonical correlation analysis (CCA). Then, shared anchors or mutual nearest neighbors (MNN) of each cell across all datasets are identified using the previously constructed CCA subspace (Haghverdi et al., 2018). This is achieved using the _"FindIntegrationAnchors"_ function provided in the _Seurat_ tool (Stuart et al., 2019). In the next step, using the identified shared anchors, _Seurat_ aligns the datasets by integrating them into a shared space and correcting the batch effects. The _"IntegrateData"_ function of the _Seurat_ tool was used to integrate the datasets (Stuart et al., 2019).

## <span style='font-size: 16px; font-weight: bold;'>Selection of the number of components or GEPs (K) in _Liger_</span>
Selecting the number of components (K) is one of the most critical steps in analyzing scRNA-seq data. _Liger_ provides several functionalities to guide users in selecting an appropriate value of K. 
The _"suggestK"_ function of the _Liger_ tool visually suggests an appropriate value of K. This function plots the median Kullback-Leibler (KL) divergence as a function of K. As K increases, the divergence increases but starts to level off once an optimal number of K is reached. A detailed explanation can be found (Welch et al., 2019).
The _"calcAgreement"_ function provided by the _Liger_ tool quantifies how much the factorization and alignment distort the geometry of the original datasets. The metric produces scores ranging from 0 to 1. A higher score indicates a more significant agreement, indicating less geometry distortion. The agreement score is typically more significant for a large value of K. In practice, the agreement score ranges between 0.2 to 0.3. Users can use this metric to identify an optimal value of K. A detailed explanation can be found (Welch et al., 2019).

The _"calcAlignment"_ function of the _Liger_ tool quantifies how well the given datasets are aligned with each other. The metric produces scores ranging from 0 to 1. A higher score indicates perfectly mixed datasets in the integrated space. Utilizing this metric, we can guide ourselves in deciding an optimal value of K. A detailed explanation can be found (Welch et al., 2019).

Relying on these metrics alone is not sufficient to select an optimal value of K. One must take the biological interpretation of the GEPs (latent factors; K) and the arrangement of cell clusters for a particular choice of K into consideration while deciding on a value of K. Determining the number of components or GEPs (K) for _Liger_ or NMF is an important but challenging step since no simple method works for all datasets and applications. Because NMF is an unsupervised, exploratory approach, K has no single "right" value. In practice, the decision on choosing the number of K is made based on a combination of biological prior knowledge and users' choice of metrics (Kotliar et al., 2019). For this reason, alongside the metrics stated above, the _Liger_ integration of the datasets was run with varying Ks or GEPs (K = 10 to 50) and considered the biological interpretability of GEPs found from various K choices to select an appropriate value of K.

## <span style='font-size: 16px; font-weight: bold;'>Calculation of replicate or group specificity score of latent factors</span>
The function _"calcDatasetSpecificity"_ of the _Liger_ tool was used to calculate each factor's dataset (replicate) specificity score (Welch et al., 2019). By default, this function calculates the specificity score between two datasets, showing whether a GEP is specific to dataset one or the other. In addition, the specificity score of the GEPs has been calculated depending on the context of the analysis. For a detailed explanation of the method, see (Welch et al., 2019).

For the integrative analysis of biological replicates (same biological tissues), the dataset-specificity score of the GEPs has been calculated for each dataset. The specificity score was calculated for all possible combinations of datasets concerning the given dataset. Then the scores were averaged across all comparison pairs to get the dataset specificity score for the given dataset against the rest.

For the comparative analysis between groups (experimental conditions, genotypes, or species), each replicate of a particular group was compared against all the replicates of the other group. Then the scores were averaged across all comparisons to get a replicate specificity score against the group. Then, the replicate specificity score of all the replicates for a particular group was averaged to get a group specificity score for the GEPs.

## <span style='font-size: 16px; font-weight: bold;'>GO enrichment analysis</span> {#GO}
Gene ontology (GO) enrichment analysis of a given set of genes for both molecular function and the biological process has been performed using a combination of two R packages, biomartr (Durinck et al., 2009) and clusterProfiler (T. Wu et al., 2021) with a p-value cutoff of 0.05. And the GO annotations of the genes, such as biological processes, cellular components, and molecular function, have been retrieved from the Ensembl databases.

## <span style='font-size: 16px; font-weight: bold;'>Inference and characterization of GEPs</span>
The shared basis matrix inferred by the _Liger_ tool shows the relative expression of genes to each GEP. Each GEP is characterized by a set of numbers describing the magnitude of association of the genes to the GEP. Genes with large values in a given GEP are the genes that are, therefore, characteristic of this GEP. We can identify sets of coexpressed genes from the basis matrix based on the relative contribution of the genes to the GEPs. Based on the GO annotation of these characteristic genes, we categorized each GEP as reflecting identity GEP (i.e., a specific tissue or cell type) or rather an activity GEP (e.g., cell cycle, stress response).

## <span style='font-size: 16px; font-weight: bold;'>Estimation of GEPs usage by the cells</span> {#gepu}
The shared coefficient matrix inferred by the _Liger_ tool shows the estimated proportions of GEP usage for each cell. The shared coefficient matrix can be used to cluster the cells based on the fraction of GEPs used by the cells. However, in this analysis, we treated this shared coefficient matrix, GEPs loadings for each cell, as a low-dimensional representation of the data, which is analogous to the principal components in PCA.

## <span style='font-size: 16px; font-weight: bold;'>Dimensionality reduction with PCA</span>
Principal component analysis (PCA) is an unsupervised linear dimensionality reduction algorithm that highlights variation and identifies hidden and potent patterns in a dataset (Jollife & Cadima, 2016). Although PCA is a dimensionality reduction method, it can be applied for visualization and noise reduction to facilitate exploratory data analysis. The principal components (PCs) serve as the input features in the _Seurat_ clustering analysis. PCA was performed on the scaled data (Section [Zero mean and unit variance scaling](#std_scaling) using the _"RunPCA"_ function of the _Seurat_ tool (Stuart et al., 2019). See (Jollife & Cadima, 2016) for a detailed explanation of the method.

## <span style='font-size: 16px; font-weight: bold;'>Dimensionality reduction with UMAP</span>
Uniform Manifold Approximation and Projection (UMAP) (Becht et al., 2019) is a nonlinear dimensionality reduction technique for visualizing high-dimensional data. UMAP was used to visualize cells in a low-dimensional space. UMAP was performed using the _"RunUMAP"_ function of the _Seurat_ tool (Stuart et al., 2019).

In the _Liger_ analysis, the shared coefficient matrix inferred by the _Liger_ tool was given as the input for UMAP to map the high-dimensional data to a two-dimensional feature space (UMAP dimensions) for visualization.

In the _Seurat_ analysis, 50 principal components inferred by the PCA method were used as input for UMAP to map the high-dimensional data to a two-dimensional feature space (UMAP dimensions) for visualization.

## <span style='font-size: 16px; font-weight: bold;'>Identification of cell clusters</span>
To identify cell clusters, we have used the Louvain method for community detection (Blondel et al., 2008), a graph-based clustering method commonly used for clustering analysis in single-cell data. Graph-based clustering technique excels at merging small clusters into broad cell classes. Clustering analysis using the _Seurat_ tool involves two steps (Stuart et al., 2019). First, a k-nearest neighbors (k-NN) graph is constructed by identifying each cell's k-nearest neighbors (k, number of neighbors). Using the constructed k-NN graph, a Shared Nearest Neighbor (SNN) Graph is constructed for the integrated data by calculating the neighborhood overlap between every cell and its k-nearest neighbors. This was achieved by utilizing the _"FindNeighbors"_ function of the _Seurat_ tool (Stuart et al., 2019). For a detailed explanation of the construction of the SNN graph, see (Haghverdi et al., 2018; Stuart et al., 2019). Finally, cell clusters are identified by a shared nearest neighbor (SNN) modularity optimization-based clustering algorithm, the Louvain algorithm. A detailed description of the clustering algorithm can be found (Blondel et al., 2008; Waltman & Jan Van Eck, 2013). The _"FindClusters"_ function of the _Seurat_ tool was used to identify cell clusters in the integrated data (Stuart et al., 2019).

In the _Liger_ analysis, the shared coefficient matrix inferred by the _Liger_ tool was given as the input for the graph-based clustering to identify cell clusters.

In the _Seurat_ analysis, 50 principal components inferred by the PCA method were used as the input features for the graph-based clustering.

## <span style='font-size: 16px; font-weight: bold;'>Identification of DEGs</span> {#ideg}
To identify genes that are significantly differentially expressed (up regulated) in specific cell types, differential expression (DE) analysis was performed using the _"FindMarkers"_ function of the _Seurat_ tool (Stuart et al., 2019). We used the non-parametric Wilcoxon rank sum test implemented in the function to identify differentially expressed genes (DEGs) for each cell cluster. See (Haynes, 2013) for a detailed method explanation. Genes detected in at least 10% of the cells in the cluster of interest (target cluster) and with an average expression fold-change of more than 0.25 compared to other clusters in log-normalized space were considered differentially expressed genes (DEGs). The output of the DE analysis contains a set of information, such as p-value, average log2 fold-change, percentage of cells expressing a gene within the target cluster (pct.1), percentage of cells expressing a gene outside of the target cluster (pct.2), and adjusted p-value.

## <span style='font-size: 16px; font-weight: bold;'>Cluster-specific markers selection</span> {#clsp}
Two selection criteria were employed to select cluster-specific marker genes from the identified DEGs for the cell clusters. First, to be a cluster-specific marker, a gene must be expressed in more than 10% of cells within the target cluster (pct.1; percentage of cells expressing the gene in the focal cluster of interest) and less than 10% of cells in the rest of the cell population (pct.2; percentage of cells expressing the gene in the rest of the cell population). This criterion was proposed by Denyer et al. (Denyer et al., 2019). Second, a gene is defined as a cluster-specific marker if the difference in detection in the target cluster (pct.1) is 30% or more than the rest of the cell population (pct.2). This criterion was proposed in this thesis. By combining the cluster-specific markers selected by these two criteria, a set of cluster-specific markers was identified for each cell cluster.

## <span style='font-size: 16px; font-weight: bold;'>Cluster-candidate biomarkers selection</span>
In this thesis, a novel candidate markers selection approach has been proposed to select a set of markers, throughout the thesis referred to as the cluster-candidate biomarkers, utilizing the characteristic GEP for a given cluster and the list of cluster-specific markers (Section [Cluster-specific markers selection](#clsp)) for a given cluster. The aim was to propose a significantly smaller set of markers for each cell cluster with significant discriminating power defined by cluster-specific higher expression signal, to differentiate a cluster from the rest of the cell clusters. Utilizing the shared coefficient matrix (inferred by the _Liger_ tool), the characteristic GEP was identified for each cell cluster. In some cases, when cells in a cluster used more than one GEP to a significant extent, we considered multiple GEPs for that cell cluster to select marker genes from the GEPs. The cluster-candidate biomarkers selection approach follows the following steps:

### <span style='font-size: 16px; font-weight: bold;'>Selection of candidate biomarkers from GEP</span>
To select candidate biomarkers, a maximum number of 10 genes, from the GEP, a sensitivity measure was employed based on the percentage of cells expressing a gene in the target cluster of interest (pct.1) and the percentage of cells expressing that gene in the rest of the cell population (pct.2). The sensitivity measure was defined as the expression signal of a gene has to be detected in 10% or more cells in the target cluster and less than 10% of cells in the rest of the cell population. To implement the sensitivity measure, we utilized the _"FindMarkers"_ function (Stuart et al., 2019) of the _Seurat_ tool to perform differential expression analysis (Section [Identification of DEGs](#ideg)) with the Wilcoxon rank sum test only for the genes of the GEP between the target cluster and the rest of the cell clusters. Genes detected in at least 10% of the cells in the target cluster and with an average expression fold-change of more than 0.25 compared to other clusters in log-normalized space were considered for the candidate biomarker selection process. If no gene passed the differential expression test, no candidate markers were reported from that GEP.

If less than 10 genes passed the differential expression test, genes that passed the test were selected as candidate biomarkers (of the GEP). The sensitivity measure was implemented if 10 or more genes passed the differential expression test. If less than 10 genes were identified based on the sensitivity measure, we selected the top 10 genes (when applicable) with the lowest pct.2 detection. If more than 10 genes were identified based on the sensitivity measure, we considered the fold-change information (calculated in log-normalized space) of the genes. First, we looked for genes with a fold-change (log2) of 1 or more. If 10 or more genes had a fold-change of 1 or more, we extracted those genes, sorted them based on pct.2 in ascending order (lowest to highest), and then picked the top 10 genes as candidate biomarkers. If less than 10 genes had a fold-change of 1 or more, we selected the genes with a fold-change of 1 or more, then we sorted the remaining genes based on the fold-change in descending order (highest to lowest) and selected the top genes from the sorted list. If no genes had a fold-change of 1, we sorted the genes based on fold-change in descending order (highest to lowest), then picked the top 10 genes as candidate biomarkers.

### <span style='font-size: 16px; font-weight: bold;'>Selection of candidate biomarkers from cluster-specific markers</span>
Genes were first categorized into two groups based on transcription factor (TF) information to select candidate biomarkers from the cluster-specific markers set (Section [Cluster-specific markers selection](#clsp)). A list of TFs for _A. thaliana_ was retrieved from the PlantTFDB (Jin et al., 2017), then using one-to-one orthologue ids of corresponding _C. hirsuta_ genes, TF information was annotated to the _C. hirsuta_ genes. The candidate biomarkers selection process was applied to both groups, TF and non-TF. To avoid duplicates in the list of candidate biomarkers, the candidate biomarkers selected from the GEP was removed from the cluster-specific markers set (when applicable). 

For the TF group, a maximum number of 10 TFs was selected as candidate biomarkers. If less than 10 or 10 TFs were present, all of them were selected as candidate biomarkers. If more than 10 TFs were present in the cluster-specific markers set, we considered the fold-change (calculated in log-normalized space) information of the TFs. First, we looked for TFs with a fold-change (log2) of 1 or more. If 10 or more TFs had a fold-change of 1 or more, we extracted those TFs, sorted them based on pct.2 in ascending order (lowest to highest), and then picked the top 10 TFs as candidate biomarkers. If less than 10 TFs had a fold-change of 1 or more, we selected the TFs with a fold-change of 1 or more, then we sorted the remaining TFs based on the fold-change in descending order (highest to lowest) and selected the top TFs from the sorted list. If no TFs had a fold-change of 1, we sorted the TFs based on fold-change in descending order (highest to lowest), then picked the top 10 TFs as candidate biomarkers. If no TFs were present in the cluster-specific markers set, no TF was reported in the cluster-candidate biomarkers set.

For the non-TF group, a maximum number of 10 genes was selected as candidate biomarkers. If less than 10 or 10 genes were present, all of them were selected as candidate biomarkers. If more than 10 genes were present in the cluster-specific markers set, we considered the genes' fold-change (calculated in log-normalized space) information. First, we looked for genes with a fold-change (log2) of 1 or more. If 10 or more genes had a fold-change of 1 or more, we extracted those genes, sorted them based on pct.2 in ascending order (lowest to highest), and then picked the top 10 genes as candidate biomarkers. If less than 10 genes had a fold-change of 1 or more, we selected the genes with a fold-change of 1 or more, then we sorted the remaining genes based on the fold-change in descending order (highest to lowest) and selected the top genes from the sorted list. If no genes had a fold-change of 1, we sorted the genes based on fold-change in descending order (highest to lowest), then picked the top 10 genes as candidate biomarkers. If no genes were present in the cluster-specific markers set, no gene was reported in the cluster-candidate biomarkers set.

A set of candidate biomarkers was selected for each cell cluster by combining the two cluster-candidate biomarkers selection steps.

## <span style='font-size: 16px; font-weight: bold;'>Module score of a gene set</span>
The _"AddModuleScore"_ function of the _Seurat_ tool (Stuart et al., 2019), which compares the expression of the genes in a given set of our interest to that of a set of control genes with expression levels, has been used to calculate the gene set activity. See (Tirosh et al., 2019) for a detailed explanation of the method.

## <span style='font-size: 16px; font-weight: bold;'>Assignment of cell type identity to the cell clusters</span>
I have used a combination of three approaches to annotate a cell cluster with a cell type identity. This novel approach for annotating cell clusters has been proposed in this thesis. First, we projected the expression levels of known cell type markers derived from the literature across the clusters. Expression of the known cell type markers helped us understand the nature of the cell clusters. Second, we investigated the shared coefficient matrix (inferred by the _Liger_ tool; see Section [Estimation of GEPs usage by the cells](#gepu)) for each cell cluster and characterized them with their most contributing GEP. GO enrichment analysis for biological process (see Section [GO enrichment analysis](#GO)) of the genes in the characteristic GEP of a cell cluster revealed cell-type related enriched GO terms which results in defining the identity of the cells in the cell clusters. Third, GO enrichment analysis of the cluster-specific markers of a cell cluster further revealed the biological processes in which these genes are involved, which results in the identification of the cellular identity of the cell clusters. We combined the findings of these three approaches to define the cell-type identity of the cell clusters.

## <span style='font-size: 16px; font-weight: bold;'>Similarity measure between clustering results</span>
To validate the clusters identified by one approach and evaluate the robustness and reliability of one clustering method, we compared the clustering result produced by a clustering method to the clustering result of a different method; the clustering result produced by the _Liger_ method and the clustering result produced by the _Seurat_ tool. We have also assessed the agreement and consistency of clustering results between two conditions, with the cell cycle effect and without the cell cycle effect. To assess agreement between two clustering results, I have used two different metrics, the V-measure, and the completeness metric.

V-measure is an external entropy-based measure that evaluates clustering performance based on the actual class label information of the data points (Rosenberg & Hirschberg, 2007). V-measure quantifies the two criteria, homogeneity, and completeness, for a clustering algorithm's result. The homogeneity criterion indicates how perfectly only the data points of a specific class are assigned to a single cluster by the clustering method. The completeness criterion shows if the clustering method has assigned all the data points of a particular class to a single cluster. Although V-measure considers the completeness metric to evaluate the clustering solution, I have calculated the completeness score separately in this thesis. A detailed explanation of the method and calculation of the score can be found (Rosenberg & Hirschberg, 2007).

We lack actual or ground truth cell-type labels for the cells for the scRNA-seq datasets of leaf tissue samples of _A. thaliana_ and _C. hirsuta_. For this reason, we treated the clustering results generated by the _Liger_ analysis as the ground truth cluster labels to compare against clustering results produced in a different condition (after removing the cell cycle effect) or clustering results produced by the _Seurat_ analysis.

## <span style='font-size: 16px; font-weight: bold;'>Statistical test for overrepresentation</span>
To determine if there is a significant association (statistically significant enrichment) between a given set of genes, such as a set of cluster-specific markers or genes in a GEP, and the presence of peaks in the entire genome (identified by a ChiP-Seq experiment; 35S::cdsRCOvenus), the Fisher's exact test for overrepresentation was performed. The _"fisher.test"_ function of the R-package stats was used to perform the Fisher's exact test. See (Sprent, 2011) for a detailed method explanation.

## <span style='font-size: 16px; font-weight: bold;'>Trajectory analysis</span>
The Slingshot tool (Street et al., 2018) was used to perform trajectory analysis for reconstructing developmental lineages. Based on the prior information required to be given as input by different trajectory inference methods, we have decided to use Slingshot to identify developmental lineages and pseudotime inference, as the Slingshot method does not require users to specify the starting cell cluster or end cell cluster of the developmental trajectory. The minimal input to the Slingshot tool is a matrix representing the cells in a low-dimensional space, such as the shared coefficient matrix or the principal components, and a vector of cluster labels of the cells. With these two inputs, Slingshot Identifies the global lineage structure by constructing a minimum spanning tree (MST) connecting the cluster centroids with the _"getLineages"_ function (Street et al., 2018). The following step constructs smooth lineages and infers pseudotime variables by fitting simultaneous principal curves with the _"getCurves"_ function.
